# Assignment 5

Turn in the assignment via Canvas.

To write legible answers you will need to be familiar with both [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) and [Latex](https://www.latex-tutorial.com/tutorials/amsmath/)

Before you turn this problem in, make sure everything runs as expected. First, restart the kernel (in the menubar, select Runtime→→Restart runtime) and then run all cells (in the menubar, select Runtime→→Run All).

Make sure you fill in any place that says "YOUR CODE HERE" or "YOUR ANSWER HERE", as well as your name below:

In [ ]:
NAME = "Juan Lee"
STUDENT_ID = ""

# Question 1

Using MDP, model the spread of COVID-19 for creating a simulator, under certain assumptions in a hypothetical situation. In the simulation, the inhabitants of a town move around as they go about their regular business. (The simulation would be similar to the one in [this link](https://www.washingtonpost.com/graphics/2020/world/corona-simulator/) if you are curious. However, the simulation itself is unrelated to modeling the MDP in this problem, and you can ignore it.)

In the simulation, there are 2 types of people in the town, a **carrier** or **non-carrier** of the Coronavirus. The virus spreads via droplets when an infected person comes in close contact (within 6 feet) of another person who has the virus, a carrier. Non-carriers cannot transmit the disease. Associated with these two types are 5 different states that a person can be in. A carrier can be in one of the following 2 states: infected (but not sick i.e. asymptomatic), or sick. A non-carrier can be in 3 states: unexposed, dead, or immune.

The states (with a distinct color representing the state in the MDP) are described as the following:


*   **Unexposed** (BLUE): The people who haven't encountered the virus at all so far.
*   **Infected** (ORANGE): People who have the virus but have no symptoms of the disease. An unexposed person has an 80% chance of getting infected on coming in contact with a person carrying the virus. An infected person may get sick with symptoms in 5 days or stay infected (contagious) for 15 days and develop immunity to the disease after that.
*   **Sick** (RED): People who have the virus and are sick with symptoms. Of those infected, 50% get sick with the disease. In 10 days, a sick person may recover completely (98%), develop lifelong immunity, or die (2%).
*   **Dead** (GRAY): People who die from the disease. 2% of the sick die.
*   **Immune** (GREEN): There are two ways of getting to this state: 1) People who got sick with COVID-19, and recovered and thereby, have developed lifelong immunity, 2) People who were infected but didn't develop symptoms and became immune once they stopped being contagious.


Lets assume that, the cost for running 1 day of simulation is 1 standard processing cycle of a specific computer. On day 1 of the simulation, a (random) person returned to the town from elsewhere and is sick with the virus. Unexposed people who come in contact with this carrier as they travel between home and work may become infected with the virus. There is an 80% chance of contracting the virus (and changing state to infected) on contact with a carrier. We assume that this is the only means by which an unexposed person may get infected by the virus, and they are infected immediately. To summarize, the following transitions between states are possible:

*   Unexposed → Infected (happens immediately on the same day after each contact with 80% chance)
*   Infected → Sick (after 5 days of getting infected with 50% chance)
*   Infected → Immune (after 15 days of getting infected with 50% chance)
*   Sick → Immune (after 10 days of getting sick with 98% chance)
*   Sick → Dead (after 10 days of getting sick with 2% chance)

a) Draw the MDP graphically
  - A good way to do this is through [Google Drawings](https://docs.google.com/drawings)
  - When you're done you can embed it in the jupyter notebook using markdown syntax \!\[alt-text\]\(url\)
  - To get the URL for your image in Google Draw goto File->Publish to the web...->Embed and copy the src portion of the html tag

b) Create the transition probabilities matrix for the MDP.

[YOUR ANSWER HERE]

a)  ![alt-text](<img src="https://docs.google.com/drawings/d/e/2PACX-1vQ4X_qUjdZr5VDHps2tjQ4C8hZLDfFtcStsCaEdQ0BHQXk3_FPHguF4zSSd3Rl3J0Khyr645mfjBccp/pub?w=960&amp;h=720">
)

The cost for running 1 day of simulation is 1 standard processing cycle of a specific computer.

b)![alt-text](<img src="https://docs.google.com/drawings/d/e/2PACX-1vQqLdIBp53t2E5GEgXV4vfOoHZH4AMKn8l7c8oQ4bR0ctCykidr2zKwMnEN0E-jWVAy_O9ANgROTJnX/pub?w=960&amp;h=720">)


# Question 2

Solve the following  MDP using both value iteration and policy iteration.

Neo is a freelance computer programmer, and his aim in life is to earn as much money as he can by writing more code. He has three possible states in his daily professional life as below.

1.   **Productive** - He can write programming codes more efficiently.
2.   **Exhausted** - He is too tired to think well to write efficient programming codes.
3.   **Fit** - He is physically and mentally in a great state to write code and solve programming problems.


When Neo is **Exhausted**, he can choose one of three actions: (1) Keep **Coding**, (2) do some physical **Exercise**, or (3) get some **Rest**.

If he chooses to do more Coding, he remains in the **Exhausted** state with the certainty of getting a +20 dollars reward. If he decides to get Rest, he has 80% of moving to the next state, Productive, and a 20% chance of staying Exhausted.
If he doesn’t want to get Rest, he may go to the gym and do some physical Exercise. This gives him a 50% chance of entering the Productive state and a 50% chance of staying Exhausted. However, he needs to pay for the gym, so this choice results in a -10 dollars reward.

When Neo becomes **Productive**, he can go back to **Coding** and write programs more efficiently. From there, he has an 80% chance of getting Exhausted again with earning a +40 dollars reward, and a 20% chance of staying Productive while earning a +30 dollar reward.
Sometimes, when he is Productive, he wants to do some physical **Exercise**. When he Exercises in this state, he enjoys it very much and gets 100% Fit physically and mentally. However, he needs to pay for it with 10 dollars.

Once Neo reaches the state **Fit**, he is fully committed to earning more money by more **Coding**. Because he is in such a great state physically and mentally, he can write programs very efficiently. In this state, he earns a +100 dollars reward, and he keeps writing code until he is Exhausted again.

Use MDP to find the best policy to maximize his earnings/rewards over time.

1. Draw the MDP graphically
  - A good way to do this is through [Google Drawings](https://docs.google.com/drawings)
  - When you're done you can embed it in the jupyter notebook using markdown syntax \!\[alt-text\]\(url\)
  - To get the URL for your image in Google Draw goto File->Publish to the web...->Embed and copy the src portion of the html tag

2. Using a discount factor of 0.86, solve the MDP using value iteration (until the values have become reasonably stable). You should start with the values set to zero. You should show both the optimal policy and the optimal values.

3. Using a discount factor of 0.86, solve the MDP using policy evaluation (until you have complete convergence). You should start with the policy that always does **Coding**.

4. Change the MDP in three different ways: by changing the discount factor, changing the transition probabilities for a single action from a single state, and by changing a reward for a single action at a single state. Each of these changes should be performed separately starting at the original MDP, resulting in three new MDPs (which you do not have to draw), each of which is different from the original MDP in a single way. In each case, the change should be so that the optimal policy changes, and you should state what the optimal policy becomes and give a short intuitive argument for this.

**You can add more cells for Code or Text as you feel the need.**

[YOUR ANSWER HERE. ADD MORE CELLS AS NEEDED.]
![alt-text](<img src="https://docs.google.com/drawings/d/e/2PACX-1vTY3_rVZ_EshZd_bw8UlH2qAPOWp19XrdgdY8yvmgy0UgaqiNMWBEdDUzvEMLuJSjSGXe-et3JZO8Io/pub?w=960&amp;h=720">)

---



In [ ]:
# Write Your Code Here. Add more cells as needed.
Opt_val = [0, 0, 0]
Opt_pol = ['', '', '']
case1 = False
case2 = False
case3 = False
loop = True
d = .86
while(loop == True):
  for i in range(3):
    if i == 0:
      if case1 == False:
        actions = [0, 0, 0]
        actions[0] = 20 + ((d) * Opt_val[0])
        actions[1] = (.5) * (-10 + (d) * Opt_val[0]) + (.5) * (-10 + (d) * Opt_val[1])
        actions[2] = (.2) * ((d) * Opt_val[0]) + (.8) * (d * Opt_val[1])
        temp = max(actions) 
        if temp == actions[0]:
          Opt_pol[0] = 'Coding' 
        elif temp == actions[1]:
          Opt_pol[0] = 'Exercise'
        elif temp == actions[2]:
          Opt_pol[0] = 'Rest'
        if abs(temp - Opt_val[0]) <= .01:
          case1 = True
        else: 
          Opt_val[0] = temp
    elif i == 1:
      if case2 == False:
        actions = [0, 0]
        actions[0] = (.8) *(40 + (d)*Opt_val[0]) + (.2) * (30 + (d*Opt_val[1]))
        actions[1] = -10 + (d*Opt_val[2])
        temp = max(actions)
        if temp == actions[0]:
          Opt_pol[1] = 'Coding' 
        elif temp == actions[1]:
          Opt_pol[1] = 'Exercise'
        
        if abs(temp - Opt_val[1]) <= .01:
          case2 = True
        else:
          Opt_val[1] = temp
    elif i == 2:
      if case3 == False:
        temp = (100 + (d * Opt_val[0]))
        Opt_pol[2] = 'Coding'
        if abs(temp - Opt_val[2]) <= .01:
          case3 = True
        else: 
          Opt_val[2] = temp
    if case1 == True and case2 == True and case3 == True:
      loop = False
  
print(Opt_val)
print(Opt_pol)


[163.7845498079534, 197.1264323200339, 240.85471283483992]
['Rest', 'Exercise', 'Coding']


Using Value Iteration:

For state Exhausted, the optimal value = 163.7845498079534 , the optimal policy = 'Rest'

For state Productive, the optimal value = 197.1264323200339, the optimal policy = 'Exercise'

For state Fit, the optimal value = 240.85471283483992 , the optimal policy = 'Coding'

In [ ]:
c = [0, 0, 0]
e = [0, 0, 0]
r = [0, 0, 0]
opt_pol = ['', '', '']
case1 = False
case2 = False
case3 = False
loop = True
while(loop == True):
  for i in range(3):
    if i == 0:
      if case1 == False:
        tempc = c[0]
        c[0] = 20 + ((.86) * c[0])
        tempe = e[0]
        e[0] = (.5) * (-10 + (.86) *e[0]) + (.5) * (-10 + (.86) * e[1])
        tempr = r[0]
        r[0] = (.2) * ((.86) * r[0]) + (.8) * (.86 * r[1])
        opt_pol[0] = 'Rest'
        if abs(tempc - c[0]) <= .01 and abs(tempe - e[0]) <= .01 and abs(tempr - r[0]) <= .01:
          case1 = True

    elif i == 1:
      if case2 == False:
        tempc = c[1]
        c[1] = (.8) *(40 + (.86)*c[0]) + (.2) * (30 + (.86*c[1]))
        tempe = e[1]
        e[1] = -10 + (.86*e[2])
        opt_pol[1] = 'Exercise'
        if abs(tempc - c[1]) <= .01 and abs(tempe - e[1]) <= .01:
          case2 = True

    elif i == 2:
      if case3 == False:
        temp = c[2]
        c[2] = (100 + (.86 * c[0]))
        opt_pol[2] = 'Coding'
        if abs(temp - c[2]) <= .01:
          case3 = True
    if case1 == True and case2 == True and case3 == True:
      loop = False

optimal = [max([c[0]+20, e[0], r[0]]),max([c[1]+20, e[1], r[1]]), max([c[2]+20, e[2], r[2]])]


print(optimal)
print(opt_pol)

[162.80106112424494, 184.54019155902756, 242.80106112424494]
['Rest', 'Exercise', 'Coding']


Using Policy Evaluation:

For state Exhausted, the optimal value = 162.80106112424494 , the optimal policy = 'Rest'

For state Productive, the optimal value = 184.54019155902756, the optimal policy = 'Exercise'

For state Fit, the optimal value = 242.80106112424494 , the optimal policy = 'Coding'

4. By changing the discount factor to 0, the only reward that matters is the instantaneous rewards so the optimal policy for all the states would be coding because coding grants the highest instant reward for each state. By changing transitional probabily of excersizing for exhauted state to have probability to return to exhausted be .8 and going to productive .2 changes the optimal policy to coding instead of rest because the guarentee of earning 20 rewards grants more value in the long run. Lastly changing the reward of coding for Productive state to be 100 for returning to exhausted state would make the optimal policy for productive to be coding instead of excersize because the value calculated in the long run would provide the maximum amount of reward for the state of Productive. 

# Question 3

## Markov Decision Process (MDP) Toolbox for Python

One useful Python module for solving these types of problems is called *MDP toolbox* (pymdptoolbox). The MDP toolbox provides classes and functions for the resolution of descrete-time Markov Decision Processes. The list of algorithms that have been implemented includes backwards induction, linear programming, policy iteration, q-learning and value iteration along with several variations. In the next cell you'll see how to load this module into a Jupyter notebook running in Colab. 

In [ ]:
!pip install pymdptoolbox

A simple example code is given below. For details about the example, check the documentation for Python MDP Toolbox (pymdptoolbox) from this [link](https://pymdptoolbox.readthedocs.io/en/latest/index.html).

In [ ]:
import numpy as np
import mdptoolbox
import mdptoolbox.example
np.random.seed(0) # Needed to get the output below
P, R = mdptoolbox.example.rand(2, 2)

pi = mdptoolbox.mdp.PolicyIteration(P, R, 0.9)
pi.run()
print(pi.policy)
print(pi.iter)



(1, 0)
1


##Solving a MDP using Python MDP Toolbox

First read the documentation for Python MDP Toolbox (pymdptoolbox) from this [link](https://pymdptoolbox.readthedocs.io/en/latest/index.html) and try to understand how it works.

Then solve the MDP from Question 2 using both value iteration and policy iteration (it is same as policy evaluation) algorithms from the Python MDP Toolbox.


In [ ]:
# Value Iteration.

states = ['Exhuasted', 'Productive', 'Fit']
actions = ['Coding', 'Exercise', 'Rest']
Prob = np.array([[[1, 0, 0],[.8, .2, 0],[1, 0, 0]], [[.5, .5, 0],[0, 0, 1], [0, 0, 1]], [[.2, .8, 0],[0, 1, 0],[0, 0, 1]]])
Reward = np.array([[[20, 0, 0],[40, 30, 0],[100, 0, 0]], [[-10, -10, 0],[0, 0, -10], [0, 0, 0]], [[0, 0, 0],[0, 0, 0], [0, 0, 0]]])

vi = mdptoolbox.mdp.ValueIteration(Prob, Reward, 0.86)
vi.run()
print(vi.policy)
print(vi.iter)
print(vi.V)

print('\nValue Iteration:')
for i in range(3):
  a = vi.policy[i]
  print('For state', states[i], 'optimal policy', actions[a], 'the optimal values', vi.V[i])


(2, 1, 0)
35
(162.9774006496814, 196.31488750860825, 240.04128246714959)

Value Iteration:
For state Exhuasted optimal policy Rest the optimal values 162.9774006496814
For state Productive optimal policy Exercise the optimal values 196.31488750860825
For state Fit optimal policy Coding the optimal values 240.04128246714959


In [ ]:
states = ['Exhuasted', 'Productive', 'Fit']
actions = ['Coding', 'Exercise', 'Rest']
Prob = np.array([[[1, 0, 0],[.8, .2, 0],[1, 0, 0]], [[.5, .5, 0],[0, 0, 1], [0, 0, 1]], [[.2, .8, 0],[0, 1, 0],[0, 0, 1]]])
Reward = np.array([[[20, 0, 0],[40, 30, 0],[100, 0, 0]], [[-10, -10, 0],[0, 0, -10], [0, 0, 0]], [[0, 0, 0],[0, 0, 0], [0, 0, 0]]])

pi = mdptoolbox.mdp.PolicyIteration(Prob, Reward, 0.86, policy0  = [0, 0 ,0])
pi.run()
print(pi.policy)
print(pi.iter)
print('\nPolicy Iteration:')
for i in range(3):
  a = pi.policy[i]
  print('For state', states[i], 'optimal policy', actions[a], 'the optimal values', pi.V[i])

(2, 1, 0)
3

Policy Iteration:
For state Exhuasted optimal policy Rest the optimal values 163.83251784711635
For state Productive optimal policy Exercise the optimal values 197.17053019972724
For state Fit optimal policy Coding the optimal values 240.89596534852006
